### PR TITLE
Added a check to VariantAnnotator that all read samples are present in the VCF

### LINKED ISSUE #6915: VariantAnnotator returns NullPointerException
Error Message from VariantAnnotator is  a NullPointerException. The user found the issue, but the error message could be improved. The issue involves the sample names from the VCF not existing in the BAM file. 

This request was created from a contribution made by Wout Megchelenbrink on October 21, 2020 12:29 UTC.

Link: [https://gatk.broadinstitute.org/hc/en-us/community/posts/360073989631-VariantAnnotator-returns-NullPointerException](https://gatk.broadinstitute.org/hc/en-us/community/posts/360073989631-VariantAnnotator-returns-NullPointerException)

\--

If you are seeing an error, please provide(REQUIRED) :  
  
**a) GATK version used:**

4.1.8.1

**b) Exact command used:**

gatk VariantAnnotator -V WES.vcf.gz -I WES.bam --output WES\_VA.vcf -A Coverage

**c) Entire error log:**

java.lang.NullPointerException  
  
at org.broadinstitute.hellbender.tools.walkers.haplotypecaller.AssemblyBasedCallerUtils.splitReadsBySample(AssemblyBasedCallerUtils.java:162)  
  
at org.broadinstitute.hellbender.tools.walkers.annotator.VariantAnnotator.makeLikelihoods(VariantAnnotator.java:244)  
  
at org.broadinstitute.hellbender.tools.walkers.annotator.VariantAnnotator.apply(VariantAnnotator.java:234)  
  
at org.broadinstitute.hellbender.engine.VariantWalker.lambda$traverse$0(VariantWalker.java:104)  
  
at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)  
  
at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)  
  
at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)  
  
at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)  
  
at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)  
  
at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)  
  
at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)  
  
at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)  
  
at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)  
  
at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)  
  
at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)  
  
at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)  
  
at org.broadinstitute.hellbender.engine.VariantWalker.traverse(VariantWalker.java:102)  
  
at org.broadinstitute.hellbender.engine.GATKTool.doWork(GATKTool.java:1049)  
  
at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:140)  
  
at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:192)  
  
at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:211)  
  
at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)  
  
at org.broadinstitute.hellbender.Main.mainEntry(Main.java:203)  
  
at org.broadinstitute.hellbender.Main.main(Main.java:289)<br><br><i>(created from <a href='https://broadinstitute.zendesk.com/agent/tickets/38357'>Zendesk ticket #38357</a>)<br>gz#38357</i>

### PR DESCRIPTION
**Pull request by jamesemery**
Fixes #6915

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotator.java [hunk 1]
```diff
@@ -196,6 +197,14 @@ public void onTraversalStart() {
         sequenceDictionary = getBestAvailableSequenceDictionary();
         variantSamples = new IndexedSampleList(samples);
 
+        // Check that the reads have compatible samples to the variants:
+        if (hasReads()) {
+            final Set<String> readsSamples = getHeaderForReads().getReadGroups().stream().map(rg -> rg.getSample()).collect(Collectors.toSet());
+            readsSamples.forEach(readSample -> {
+                if (!samples.contains(readSample))
+                    throw new UserException(String.format("Reads sample '%s' from readgroups tags does not match any sample in the variant genotypes", readSample));
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotator.java:205]: This checks whether the samples in the bam are present in the vcf. But the original bug report in https://github.com/broadinstitute/gatk/issues/6915 was (according to @gbrandt6 ) triggered by the reverse case where samples in the vcf don't exist in the bam.

**Review by droazen — CHANGES_REQUESTED** [review]
@jamesemery I think the check you added here might be the reverse of the one required

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotator.java (hunk 1 above)
Comment (droazen) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotator.java:205]: Upon further investigation, @jamesemery and I confirmed that this patch is correct, and the original bug report was backwards. The NPE is triggered when there are samples in the bam not present in the vcf.

**Review by droazen — APPROVED** [review]

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotator.java (hunk 1 above)
Comment (gbrandt6) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotator.java:205]: Thanks for looking into it @droazen @jamesemery I checked again and it looks like the user said there were no samples in common.

Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/VariantAnnotator.java:205]: If you look at the version that user was running the NPE is actually occurring on this line:   
`returnMap.get(ReadUtils.getSampleName(read, header)).add(read); `

Which means that the problem is caused when the returnMap (which gets generated right above from the VCF samples list) doesn't contain a sample that is present in one of the reads at that site. It sounded like the user had accentually matched up the wrong bam with the wrong VCF and thats why they got the crash. Technically what they said is accurate but not exactly what caused the bug, it was the reverse of what they reported as the bug. Given that multisample bams exist perhaps there is something more robust that can be done to fix the bug if this ever comes up again.

**Pull request merged**